pallets/reward-distribution/src/lib.rs

```diff
@@ -99,5 +156,188 @@ pub mod pallet {
 			Self::deposit_event(Event::<T>::RewardPerBlockAdapted(new_reward_per_block));
 			Ok(())
 		}
+
+		#[pallet::call_index(1)]
+		#[pallet::weight(<T as Config>::WeightInfo::collect_reward())]
+		#[transactional]
+		pub fn collect_reward(
+			origin: OriginFor<T>,
+			vault_id: DefaultVaultId<T>,
+			reward_currency_id: CurrencyId<T>,
+			index: Option<T::Index>,
+		) -> DispatchResult {
+			//distribute reward from reward pool to staking pallet store
+			let caller = ensure_signed(origin)?;
+			let reward = ext::pooled_rewards::withdraw_reward::<T>(
+				&vault_id.collateral_currency(),
+				&vault_id,
+				reward_currency_id,
+			)?;
+			ext::staking::distribute_reward::<T>(&vault_id, reward, reward_currency_id)?;
+
+			//withdraw the reward for specific nominator
+			let caller_rewards =
+				ext::staking::withdraw_reward::<T>(&vault_id, &caller, index, reward_currency_id)?;
+
+			if caller_rewards == (BalanceOf::<T>::zero()) {
+				return Err(Error::<T>::NoRewardsForAccount.into())
+			}
+
+			//transfer rewards
+			Self::transfer_reward(reward_currency_id, caller_rewards, caller)
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	fn transfer_reward(
+		reward_currency_id: CurrencyId<T>,
+		reward: BalanceOf<T>,
+		beneficiary: T::AccountId,
+	) -> DispatchResult {
+		let native_currency_id = T::GetNativeCurrencyId::get();
```

Comment:
Here the rewards will be minted if the fee account does not have enough to pay. We thought that it may be useful since this will also allow for the possibility of charging fees in native currency (something right now is not done) but this will allow it to do so, albeit with the downside of mixing the two systems.

Reply:
I missed this part in the code. It probably works correctly but this is quite a big (implicit) assumption that we can do this – the assumption being that any native fee not in the fee account still needs to be minted.

This feels brittle to me, at least when this assumption is so implicitly hidden here at this code. It could also lead to a particularly serious attack if there is a bug somewhere else in the code and there are way more fees minted than intended.

I think it is cleaner and safer to keep track of native fees that need to be minted in an extra storage field. Thoughts?

Reply:
This check is now done [here](https://github.com/pendulum-chain/spacewalk/pull/415/commits/05b2bed97777144de080afcc1c3e5196feff911b#diff-2f743609293dfd3b50158de75d67dabc3018b7fa2abe78bed8fd45a783aa3975R218)